File: Iced.UnitTests/Intel/DecoderTests/DecoderTestCase.cs
/*
Copyright (C) 2018-2019 de4dot@gmail.com

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
"Software"), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
*/

using System;
using Iced.Intel;

namespace Iced.UnitTests.Intel.DecoderTests {
	sealed class DecoderTestCase {
		public int LineNumber;
		public DecoderOptions DecoderOptions;
		public int Bitness;
		public string HexBytes;
		public string EncodedHexBytes;
		public Code Code;
		public int OpCount;
		public bool ZeroingMasking;
		public bool SuppressAllExceptions;
		public bool IsBroadcast;
		public bool HasXacquirePrefix;
		public bool HasXreleasePrefix;
		public bool HasRepePrefix;
		public bool HasRepnePrefix;
		public bool HasLockPrefix;
		public int VsibBitness;
		public Register OpMask;
		public RoundingControl RoundingControl;
		public OpKind Op0Kind, Op1Kind, Op2Kind, Op3Kind, Op4Kind;
		public Register SegmentPrefix;
		public Register MemorySegment;
		public Register MemoryBase;
		public Register MemoryIndex;
		public int MemoryDisplSize;
		public MemorySize MemorySize;
		public int MemoryIndexScale;
		public uint MemoryDisplacement;
		public ulong Immediate;
		public byte Immediate_2nd;
		public ulong MemoryAddress64;
		public ulong NearBranch;
		public uint FarBranch;
		public ushort FarBranchSelector;
		public Register Op0Register, Op1Register, Op2Register, Op3Register, Op4Register;
		public ConstantOffsets ConstantOffsets;

		public OpKind GetOpKind(int operand) {
			switch (operand) {
			case 0: return Op0Kind;
			case 1: return Op1Kind;
			case 2: return Op2Kind;
			case 3: return Op3Kind;
			case 4: return Op4Kind;
			default: throw new ArgumentOutOfRangeException(nameof(operand));
			}
		}

		public Register GetOpRegister(int operand) {
			switch (operand) {
			case 0: return Op0Register;
			case 1: return Op1Register;
			case 2: return Op2Register;
			case 3: return Op3Register;
			case 4: return Op4Register;
			default: throw new ArgumentOutOfRangeException(nameof(operand));
			}
		}
	}
}

File: src/rust/iced-x86/src/decoder/tests/decoder_test_case.rs
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

use crate::iced_constants::IcedConstants;
use crate::*;
use alloc::string::String;

#[cfg(feature = "mvex")]
#[derive(Default)]
pub(crate) struct MvexDecoderInfo {
	pub(crate) eviction_hint: bool,
	pub(crate) reg_mem_conv: MvexRegMemConv,
}

#[derive(Default)]
pub(crate) struct DecoderTestCase {
	#[allow(dead_code)]
	pub(crate) line_number: u32,
	pub(crate) test_options: u32,
	pub(crate) decoder_error: DecoderError,
	pub(crate) decoder_options: u32,
	pub(crate) bitness: u32,
	pub(crate) hex_bytes: String,
	pub(crate) ip: u64,
	pub(crate) encoded_hex_bytes: String,
	pub(crate) code: Code,
	pub(crate) mnemonic: Mnemonic,
	pub(crate) op_count: u32,
	pub(crate) zeroing_masking: bool,
	pub(crate) suppress_all_exceptions: bool,
	pub(crate) is_broadcast: bool,
	pub(crate) has_xacquire_prefix: bool,
	pub(crate) has_xrelease_prefix: bool,
	pub(crate) has_repe_prefix: bool,
	pub(crate) has_repne_prefix: bool,
	pub(crate) has_lock_prefix: bool,
	pub(crate) vsib_bitness: u32,
	pub(crate) op_mask: Register,
	pub(crate) rounding_control: RoundingControl,
	pub(crate) op_kinds: [OpKind; IcedConstants::MAX_OP_COUNT],
	pub(crate) segment_prefix: Register,
	pub(crate) memory_segment: Register,
	pub(crate) memory_base: Register,
	pub(crate) memory_index: Register,
	pub(crate) memory_displ_size: u32,
	pub(crate) memory_size: MemorySize,
	pub(crate) memory_index_scale: u32,
	pub(crate) memory_displacement: u64,
	pub(crate) immediate: u64,
	pub(crate) immediate_2nd: u8,
	pub(crate) near_branch: u64,
	pub(crate) far_branch: u32,
	pub(crate) far_branch_selector: u16,
	pub(crate) op_registers: [Register; IcedConstants::MAX_OP_COUNT],
	pub(crate) constant_offsets: ConstantOffsets,
	#[cfg(feature = "mvex")]
	pub(crate) mvex: MvexDecoderInfo,
}
